### ISSUE TITLE
Grid does not resize properly when adding text from code behind, and other bugs.

### ISSUE THREAD

**Issue by lenonk** [AvaloniaUI/Avalonia#3228]
I was looking for a way to emulate MessageBox.ShowMessage(), and I found an SO question where Kekekeks posted some code as the answer.  Two bugs were mentioned in that post, which I verified.  

First, the entire content of the window resulting from a call to ShowMessage() seems to be drawn on the left side, leaving a black box on the right side of the window.  This can be "fixed" by resizing the window, or by minimizing and restoring it.  This can easily be reproduced via Kekekeks code [here.](https://stackoverflow.com/questions/55706291/how-to-show-a-message-box-in-avaloniaui-beta)

Secondly, Window.CanResize == false, but the application icon, the minimize icon, and the maximize icon are all available, and work.  I know you guys already know about this one, and may have an active bug for it already.

I've also managed to hit two additional bugs myself.

I wasn't happy with the look of the message box created by the posted code, so I started working with it myself.  I managed to "fix" the layout issue by using a Grid in the xaml instead of an outer Stackpanel.  But now I have the following issues:  

There is a TextBlock in row 0 of the Grid, and a StackPanel in row 1.  TextWrapping on the TextBlock is set to "Wrap".  There are two different behaviors that I've noticed:

1) With RowDefinitions="auto, auto"
When the text is set from the code, if it should wrap, the height of the Grid (and window) do not change, and the text does not wrap.  If you resize the messagebox, the text appears and pushes the bottom row of the grid down at least partially off the bottom edge of the window.

What actually happens when text is wrapped in this case:
![sizing_issue_actual](https://user-images.githubusercontent.com/1181733/68341721-a9205780-00ae-11ea-87cf-06495baa37c5.png)

What I expect to happen (I just resized the window from the first pic):
![sizing_issue_expected](https://user-images.githubusercontent.com/1181733/68341733-ae7da200-00ae-11ea-93a0-6bed962ada75.png)

2) With RowDefinitions="*, auto"
The text appears to wrap properly, but is obscured by some otherwise invisible element that exists immediately above the StackPanel in Row 1.  Here's an image of that:

![partially hidden text](https://user-images.githubusercontent.com/1181733/68344899-bc82f100-00b5-11ea-8864-02ac0c67b655.png)
Note:  The bottom line of text was originally completely obscured, but after resizing the height by a few pixels the above image emerges.  If I had to guess, this is the 25px margin on the bottom of the TextBlock that does not get reset when the text wraps.

Also, in the xaml below you'll notice a Border element in the same grid coordinates as the StackPanel.  If HorizontalAlignment is absent, or set to "Left" on the StackPanel, everything works as expected.  Buttons are drawn on the left side of the window, and the entire bottom row of the Grid is the same color.  However. if HorizontalAlignment is set to "Right" on the StackPanel, the buttons are drawn to the right, but the Background color is only respected behind the buttons.

What actually happens without the Border tag in the xaml below:
![stackpanel_background_issue_without_border](https://user-images.githubusercontent.com/1181733/68341736-b2112900-00ae-11ea-9e0f-ced88b652f29.png)

What I expect to happen:
![stackpanel_background_issue_with_border](https://user-images.githubusercontent.com/1181733/68341741-b4738300-00ae-11ea-8e18-c6125a3962b8.png)

Here is the code I've ended up with.  The only issue I've been unable to work around is the messagebox not resizing properly when text that wraps is added:

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="405" d:DesignHeight="200" MaxWidth="405"
        x:Class="MsgBox.MessageBox" SizeToContent="WidthAndHeight" CanResize="True">
    <Grid Name="Grid" 
        ColumnDefinitions="*"
        RowDefinitions="auto, auto">
        <TextBlock 
            Margin="10,25,10,25" Grid.Row="0" Grid.Column="0" 
            TextWrapping="Wrap" 
            HorizontalAlignment="Left" 
            Name="Text" />
        <Border Grid.Row="1" Grid.Column = "0" Background="{StaticResource ThemeControlHighBrush}" />
        <StackPanel Grid.Row="1" Grid.Column="0" 
                    Background="{StaticResource ThemeControlHighBrush}" 
                    Orientation="Horizontal" 
                    HorizontalAlignment="Right"
                    Name="Buttons"> 
            <StackPanel.Styles>
                <Style Selector="Button">
                    <Setter Property="Margin" Value="2,5,2,5"/>
                </Style>
            </StackPanel.Styles>
    </StackPanel>
  </Grid>
</Window>
```

```c#
using System;
using System.Reflection;
using System.Threading.Tasks;
using Avalonia.Controls;
using Avalonia.Interactivity;
using Avalonia.Markup.Xaml;

namespace MsgBox {
    class MessageBox : Window {
        public enum MessageBoxButtons {
            Ok,
            OkCancel,
            YesNo,
            YesNoCancel
        }

        public enum MessageBoxResult {
            Ok,
            Cancel,
            Yes,
            No
        }

        public MessageBox() {
            AvaloniaXamlLoader.Load(this);
        }

        public static void Show(string text) {
            Show(null, text, "", MessageBoxButtons.Ok);
        }

        public static Task<MessageBoxResult> Show(Window parent, string text, string title, MessageBoxButtons buttons) {
            var msgbox = new MessageBox() {
                Title = title
            };

            msgbox.FindControl<TextBlock>("Text").Text = text;

            var buttonPanel = msgbox.FindControl<StackPanel>("Buttons");

            var res = MessageBoxResult.Ok;

            void AddButton(string caption, MessageBoxResult r, bool def = false) {
                var btn = new Button { Content = caption };
                btn.Click += (_, __) => {
                    res = r;
                    msgbox.Close();
                };
                buttonPanel.Children.Add(btn);
                if (def)
                    res = r;
                btn.HorizontalAlignment = Avalonia.Layout.HorizontalAlignment.Right;
            }

            if (buttons == MessageBoxButtons.Ok || buttons == MessageBoxButtons.OkCancel)
                AddButton("Ok", MessageBoxResult.Ok, true);
            if (buttons == MessageBoxButtons.YesNo || buttons == MessageBoxButtons.YesNoCancel) {
                AddButton("Yes", MessageBoxResult.Yes);
                AddButton("No", MessageBoxResult.No, true);
            }

            if (buttons == MessageBoxButtons.OkCancel || buttons == MessageBoxButtons.YesNoCancel)
                AddButton("Cancel", MessageBoxResult.Cancel, true);

            var tcs = new TaskCompletionSource<MessageBoxResult>();
            msgbox.Closed += delegate { tcs.TrySetResult(res); };
            if (parent != null)
                msgbox.ShowDialog(parent);
            else msgbox.Show();

            return tcs.Task;
        }
    }
}

**Comment by kekekeks** [comment]
Are you using 0.8.3 or 0.9 preview builds? 0.9 has Grid implementation ported from WPF.

**Comment by lenonk** [comment]
8.3.  Last time I tried to upgrade to 0.9, my project wouldn't build.

**Comment by grokys** [comment]
Yeah, these problems should be fixed in 0.9 afaik. If you open another issue regarding your build errors there we'll try to help. Note that it's probably better to use `0.9.0-preview4` instead of `preview5` because we accidentally broke `TextBox` on `preview5`.

**Comment by lenonk** [comment]
So, iirc, my build errors were centered around this hacky BS:
'''c#
            // Hack around DataGrid's SelectionChanged event being inaccesible from Avalonia XAML
            DataGrid _grid = Avalonia.Application.Current.MainWindow.FindControl<DataGrid>("instanceGrid");
            if (_grid != null) {
                // Just in case this ever gets called twice...
                _grid.SelectionChanged -= OnRowClicked;
                _grid.SelectionChanged += OnRowClicked;
            } 
            else {
                MessageBox.Show("Unable to find grid control!");
            }'''

Because Avalonia.Application.Current.MainWindow was removed.  I have to do that crap because the SelectionChanged event on DataGrid doesn't work.  The error is:

Severity	Code	Description	Project	File	Line	Suppression State	Suppression State
Error	XAMLIL	Unable to find suitable setter or adder for property SelectionChanged of type Avalonia.Controls.DataGrid:Avalonia.Controls.DataGrid for argument System.Runtime:System.String, available setter parameter lists are:
System.EventHandler`1<Avalonia.Controls.SelectionChangedEventArgs> (line 37 position 11) Line 37, position 11.	

Xaml looks like this:
```xaml
        <DataGrid Name="instanceGrid"
          VerticalAlignment="Top"
          HorizontalAlignment="Left"
          IsReadOnly="True"
          AutoGenerateColumns="True"
          SelectionChanged="OnRowClicked"
          SelectedItem="{Binding SInstance, Mode=OneWayToSource}"
          Items="{Binding MyProp}">	
```

Event Handler looks like this:
```c#
        public void OnRowClicked(object sender, SelectionChangedEventArgs e) {
            e.Handled = true;
        }
```

So am I doing something wrong here or do I need to file a bug report about the SelectionChanged event not working?

**Comment by lenonk** [comment]
I commented out the parts of my app that wouldn't compile with 0.9 preview 4, and unfortunately, none of the two issues that I reported above are fixed.  The text still does not wrap properly, regardless of RowDefinitions="auto, auto" or "*, auto".  The background is still not drawn properly without the <Border> tag, and all the icons are still available, though I'm not sure you were claiming that one had been fixed.
